22 спринт, Паничев

AlexPanic opened the pull request.

Commit:
Начинаем ДЗ по 22 спринту..

Commit:
плейлисты создаются и отображаются

Commit:
все задания выполнил

Commit:
все задания выполнил

Grigoriy-M commented on app/src/main/java/com/example/playlistmaker/data/FavoritesRepositoryImpl.kt:
```diff
@@ -14,17 +14,11 @@ class FavoritesRepositoryImpl(
 ) : FavoritesRepository {
-    override fun getFavorites(): Flow<List<Track>> = flow{
-        appDatabase.favoritesDao().getFavorites().collect{
+    override fun getFavorites(): Flow<List<Track>> = flow {
```
Это можно сделать гораздо проще, например так:
![image](https://github.com/AlexPanic/PlaylistMaker/assets/151455241/5d58e170-323c-40d4-9309-889d121f83cc)

Grigoriy-M commented on app/src/main/java/com/example/playlistmaker/data/PlaylistsRepositoryImpl.kt:
```diff
@@ -0,0 +1,55 @@
+    private val playlistsConverter: PlaylistsConverter,
+    private val tracksConverter: TracksConverter,
+) : PlaylistsRepository {
+    override fun getPlaylists(): Flow<List<Playlist>> = flow {
```
Чтобы трансформировать один тип flow в другой используй метод map.

Grigoriy-M commented on app/src/main/java/com/example/playlistmaker/ui/RootActivity.kt:
```diff
@@ -24,6 +25,18 @@ class RootActivity : AppCompatActivity() {
+        navController.addOnDestinationChangedListener { _, destination, _ ->
+            when (destination.id) {
+                R.id.playlistAddFragment -> {
+                    bottomNavigationView.visibility = View.GONE
```
![view](https://github.com/AlexPanic/PlaylistMaker/assets/151455241/5c66b143-c013-4771-9707-30128a7e5f82)

Grigoriy-M commented on app/src/main/java/com/example/playlistmaker/ui/common/Helper.kt:
```diff
@@ -5,24 +5,45 @@ import android.util.TypedValue
             }
         }
+
+        fun getWordFormDependingOnNumber(
```
Не рекомендую делать такие "велосипеды". Проблема уже давно решена в Android SDK
Можно почитать про Ресурсы plurals
https://devcolibri.com/plural-ресурсы-множественное-число/
https://developer.alexanderklimov.ru/android/theory/plurals.php

Grigoriy-M commented on app/src/main/java/com/example/playlistmaker/ui/player/activity/PlayerActivity.kt:
```diff
@@ -54,20 +68,109 @@ class PlayerActivity : AppCompatActivity() {
+            PlaylistsAdapter(
+                // адаптер плейлистов с шаблоном для экрана плеера
+                layout = R.layout.list_item_playlist_on_player_screen,
+                clickListener = object : PlaylistsAdapter.PlaylistClickListener {
```
Можно переписать так
clickListener = { playlist -> viewModel.addTrackToPlaylist(track, playlist) })
Если использовать fun interface, например так:
    fun interface PlaylistClickListener {
        fun onPlaylistClick(playlist: Playlist)
    }

Grigoriy-M commented on app/src/main/java/com/example/playlistmaker/ui/playlist_add/PlaylistAddFragment.kt:
```diff
@@ -0,0 +1,182 @@
+            when (it) {
+                is PlaylistAddState.Added -> {
+                    showToast(
+                        getString(R.string.playlist) + " " + binding.newPlaylistName.text
```
Используй getString с параметрами.
https://stackoverflow.com/questions/2397613/are-parameters-in-strings-xml-possible

Grigoriy-M commented on app/src/main/java/com/example/playlistmaker/ui/playlists/PlaylistsViewHolder.kt:
```diff
@@ -0,0 +1,48 @@
+        name.text = item.name
+        with(itemView.context) {
+            tracksCount.text =
+                item.tracksCount.toString() + " " + Helper.getWordFormDependingOnNumber(
```
Этот код можно сильно упростить, если решать проблему с помощью Ресурсы plurals
https://developer.alexanderklimov.ru/android/theory/plurals.php

Commit:
исправил: новый плейлист без картинки показывался с картинкой предыдущего плейлиста

Commit:
исправил: кнопка создать не перекрывает инпут описания (проверил на Pixel2 с разрешением 1920x1080)

Commit:
исправил: сообщение об ошибке сети, хотя сеть включена.

Было вот что:
поиск по слову basta (находит), потом быстро все стираешь и запускается поиск как будто по пустой строке, а на самом деле по одной букве b. А это особый случай поиск выдает статус 500. Его, а также 404 интепретирую как ничего не найдено.

+Добавил остановку задания на поиск в модели, если стерто поле для поиска (иначе показывалась история, а потом в течение 2 сек загружался поиск по одной букве)

Commit:
fix clear text button

Commit:
Исправил: тема в приложении точно такая, какую ставили в последний раз (сохраняется)

Commit:
minorfix, поигрался с иконкой приложения

Commit:
применил замечание ревьюера в getFavorites(), getPlaylists(), сделал рефакторинг кнопок исправил интересный баг: при смене темы приложения в течение сеанса фон кнопки в избранное оставался от предыдущей темый, т.к. устанавливался каждый раз динамически от applicationContext - поставил this

Commit:
еще немного рефакторинга кнопок и цветов

Commit:
применил замечание ревьюера
